### PR TITLE
Update optimizer.py

### PR DESCRIPTION
**Pull request by szha**

### CODE REVIEW

File: python/mxnet/optimizer.py [hunk 1]
```diff
@@ -979,8 +979,9 @@ def sync_state_context(self, state, context):
     def set_states(self, states):
         """Sets updater states."""
         states = pickle.loads(states)
-        if isinstance(states, tuple) and len(states) == 2:
-            self.states, self.optimizer = states
+        if isinstance(states, dict):
```
Comment (piiswrong) [inline python/mxnet/optimizer.py]: ?

Comment (szha) [reply python/mxnet/optimizer.py]: self.states can be length 2 tuple so this is a bug fix.

Comment (piiswrong) [reply python/mxnet/optimizer.py]: But states are previously a dictionary. This is not backward compatible

Comment (piiswrong) [reply python/mxnet/optimizer.py]: why would it be a tuple?

Comment (szha) [reply python/mxnet/optimizer.py]: I missed this myself, it's here: https://github.com/apache/incubator-mxnet/blob/master/python/mxnet/optimizer.py#L375

Comment (szha) [reply python/mxnet/optimizer.py]: List then?

Comment (szha) [reply python/mxnet/optimizer.py]: I was confused. `self.states` is actually the dictionary in Updater, so the optimizer's tuple shouldn't be a problem.

**Comment by szha** [comment]
Closing for now. A more general refactor on entity serialization in mxnet may be needed.

File: python/mxnet/optimizer.py (hunk 1 above)
Comment (ptrendx) [reply python/mxnet/optimizer.py]: Ohhhh. Ok, I got confused as well - any particular state may be a tuple, but states is dictionary of them. Sorry for that.

**Pull request closed without merging**